### ISSUE TITLE
Auto tune indexing config

### ISSUE THREAD

**Issue by icefury71** [apache/incubator-pinot#6254]
Make it easier for people to create a Pinot table which can execute queries in a performant way. Usually this means the Pinot user has to configure things such as inverted indexes. We can do this automatically. 

For instance, users can specify something like:

```
     "tableIndexConfig": {
        “mode”: “auto”
        "streamConfigs" : {
           "streamType": "kafka",
           ...
         }
      }
```

which can be automatically replaced with the necessary config.

The implementation can be done in 2 parts:
1) Have a pluggable interface for defining such customizations
2) Iterate on different types of such auto-tuning implementations.

**Comment by mcvsubbu** [comment]
Do you anticipate different auto-tuning customizations? In that case, is it better to have a config like:
`autoTuningStrategy: HighThroughput`, or `autoTuningStrategy: LowCost` or whatever.

**Comment by icefury71** [comment]
> Do you anticipate different auto-tuning customizations? In that case, is it better to have a config like:
> `autoTuningStrategy: HighThroughput`, or `autoTuningStrategy: LowCost` or whatever.

Yeah most likely will need different kinds of tuning. I'm not entirely sure we can encapsulate all those in a few keywords. For instance, indexing config is one subset. The other things we could tune is segment size / rows which could need another type of implementation. Of course, we can revisit this once we know more. For now, I thought the best approach is to leave this open.

**Comment by mcvsubbu** [comment]
> > Do you anticipate different auto-tuning customizations? In that case, is it better to have a config like:
> > `autoTuningStrategy: HighThroughput`, or `autoTuningStrategy: LowCost` or whatever.
> 
> Yeah most likely will need different kinds of tuning. I'm not entirely sure we can encapsulate all those in a few keywords. For instance, indexing config is one subset. The other things we could tune is segment size / rows which could need another type of implementation. Of course, we can revisit this once we know more. For now, I thought the best approach is to leave this open.

Do you anticipate exactly one auto-tuning customization to be used? I was just thinking of specifying a class name or a list of classes to apply the tuning. Of course, the interface would still be there, and we can have a default implementation that does a few automatic settings.

**Comment by mcvsubbu** [comment]
Another question. Do you expect that the 'auto' mode goes back and changes the table config with the settings (e.g. inverted indices etc.)? In other words, if we displayed the table config, will it show the automatically populated settings? If there is an interface and there can be multiple implementations, should it also not show which implemntation was used to arrive at the optimal config?

**Comment by icefury71** [comment]
> > > Do you anticipate different auto-tuning customizations? In that case, is it better to have a config like:
> > > `autoTuningStrategy: HighThroughput`, or `autoTuningStrategy: LowCost` or whatever.
> > 
> > 
> > Yeah most likely will need different kinds of tuning. I'm not entirely sure we can encapsulate all those in a few keywords. For instance, indexing config is one subset. The other things we could tune is segment size / rows which could need another type of implementation. Of course, we can revisit this once we know more. For now, I thought the best approach is to leave this open.
> 
> Do you anticipate exactly one auto-tuning customization to be used? I was just thinking of specifying a class name or a list of classes to apply the tuning. Of course, the interface would still be there, and we can have a default implementation that does a few automatic settings.

Good point ! It's possible that we have a chain of resolvers (similar to ingestion side transforms). We can consider this in the future. For now, I'm thinking of exactly one resolver (since we're still experimenting).

**Comment by icefury71** [comment]
> Another question. Do you expect that the 'auto' mode goes back and changes the table config with the settings (e.g. inverted indices etc.)? In other words, if we displayed the table config, will it show the automatically populated settings? 

It's technically upto the implementation. But yes, I was thinking that the resolver will populate the recommended config and persist the changes. The next time you query the table config - you should be able to see the "resolved" config.

> If there is an interface and there can be multiple implementations, should it also not show which implemntation was used to arrive at the optimal config?

Currently I'm thinking of exactly one resolver in the controller config. With this in mind, probably not important to show this.

**Comment by siddharthteotia** [comment]
Hi @icefury71 , config recommendation engine was recently implemented and that is aimed at auto-tuning as well. Did you take a look at that?

**Comment by siddharthteotia** [comment]
I don't think we should implement another auto-tuning engine. WE should enhance the existing recommendation engine

**Comment by kishoreg** [comment]
@siddharthteotia as far as I know, the recommendation engine you are referring to is a tool that recommends a config - it does not change anything in the table config. This PR is more about having the ability to plugin the engine you are referring to.

@icefury71 I like what Subbu is suggesting, we can actually have named autoTuningStrategy or a filter chain as you suggested. By default it can be a no-op. This allows multiple strategies to be registered at startup or by default and users can pick the right strategy on a per-table basis.

**Comment by icefury71** [comment]
> @siddharthteotia as far as I know, the recommendation engine you are referring to is a tool that recommends a config - it does not change anything in the table config. This PR is more about having the ability to plugin the engine you are referring to.
Yes - the goal is to have a plugin interface.

> @icefury71 I like what Subbu is suggesting, we can actually have named autoTuningStrategy or a filter chain as you suggested. By default it can be a no-op. This allows multiple strategies to be registered at startup or by default and users can pick the right strategy on a per-table basis.

Sure lemme make that change.

**Comment by icefury71** [comment]
I updated the diff with a way to register "named" TableConfigTuner methods. Also added an annotation type to automatically discover such tuner methods in the classpath (based on recommendation from @kishoreg )
